OpenRA.Mods.D2k/Traits/Sandworm.cs

```diff
@@ -99,33 +100,35 @@ void RescanForTargets(Actor self)
 				if (!actor.IsInWorld)
 					continue;
 
-				// TODO: Test if we really want to ignore actors that are on rock
 				if (!mobile.Value.CanEnterCell(actor.Location, null, false))
 					continue;
 
-				var noise = actor.TraitOrDefault<AttractsWorms>();
-				if (noise == null)
+				var noiseTraits = actor.TraitsImplementing<AttractsWorms>().Where(x => !x.IsTraitDisabled);
```

Comment:
The way the noise calculation is abstracted doesn't feel right.

I suggest implementing `public WVec AttractionAtPosition(WPos pos)` on `AttractsWorms` which returns `WVec.Zero` (trait disabled) or the attraction vector component that you are currently calculating below.  The code here then simplifies to:

```csharp
var centerPosition = self.CenterPosition;
Func<Actor, bool> isValidTarget = a =>
{
	if (!a.IsInWorld)
		return false;

	if (!a.HasTrait<AttractsWorms>())
		return false;

	var distance = (a.CenterPosition - centerPosition).HorizontalLengthSquared;
	if (distance > Info.IgnoreNoiseAttackRange * Info.IgnoreNoiseAttackRange)
		return false;

	return mobile.Value.CanEnterCell(a.Location, null, false);
};

var actorsInRange = self.World.FindActorsInCircle(self.CenterPosition, Info.MaxSearchRadius)
	.Where(isValidTarget).SelectMany(a => a.TraitsImplementing<AttractsWorms>());

var noiseDirection = actorsInRange.Aggregate(WVec.Zero, (a, b) => a + b.AttractionAtPosition(centerPosition));
```